### PR TITLE
Motion control node

### PR DESCRIPTION
**Pull request by kylestach**
## Description
Clean up the motion control node in preparation for moving to a pub-sub.

## Associated Issue
Issue #1424 

## Design Documents
[ROS Roadmap](https://docs.google.com/document/d/1wix-ZK9NPBY0zC8T9LyUihT1ONubXMKxHvb8KYmJkNQ/edit#)

## Steps to test
Motion control should experience no regressions. Run adaptive formation to test.

### CODE REVIEW

File: soccer/motion/MotionControl.cpp [hunk 1]
```diff
@@ -35,244 +30,150 @@ void MotionControl::createConfiguration(Configuration* cfg) {
...
-            angleVel > 0 ? minEffectiveAngularSpeed : -minEffectiveAngularSpeed;
+    // Debug drawing
+    {
+        if (at_end) {
+            _drawer->drawCircle(maybe_target->motion.pos, .15, Qt::red,
+                                "Planning");
+        } else if (maybe_target) {
+            _drawer->drawCircle(maybe_target->motion.pos, .15, Qt::green,
+                                "Planning");
+        }
```
Comment (JNeiger) [inline soccer/motion/MotionControl.cpp:129]: You lost the velocity debug drawer

File: soccer/motion/MotionControl.cpp [hunk 2]
```diff
@@ -35,244 +30,150 @@ void MotionControl::createConfiguration(Configuration* cfg) {
...
-    //} else {
-    if (optTarget) {
-        if (optTarget->angle) {
-            if (optTarget->angle->angle) {
-                targetAngleFinal = *optTarget->angle->angle;
-            }
-        }
-    }
-    //}
+    // TODO: Reset and return if robot is joystick-controlled.
```
Comment (JNeiger) [inline soccer/motion/MotionControl.cpp]: I would make sure this is added before actually merging. It would be a real surprise to whoever tests next if you dont lol

File: soccer/motion/MotionControlNode.cpp [hunk 3]
```diff
@@ -1,28 +1,30 @@
...
+}
 
+void MotionControlNode::runMotion(
+    const WorldState& world_state, const GameState& game_state,
+    const std::array<Planning::AngleFunctionPath, Num_Shells>& paths,
+    std::array<MotionSetpoint, Num_Shells>* setpoints) {
+    bool force_stop = game_state.state == GameState::State::Halt;
+    for (int i = 0; i < Num_Shells; i++) {
+        MotionControl& controller = _controllers[i];
+        MotionSetpoint* setpoint = &(*setpoints)[i];
```
Comment (JNeiger) [inline soccer/motion/MotionControlNode.cpp:24]: This looks a little jank from a syntax pov, but there's not really a way to fix it :/

File: logging/CMakeLists.txt [hunk 4]
```diff
@@ -8,3 +7,7 @@ target_link_libraries(simple_logger common rc-fshare)
 
 add_executable(log_converter log_converter.cpp)
 target_link_libraries(log_converter common rc-fshare)
+
+# We need XML because Context includes configurables,
+# which uses XML to represent the tree.
+qt5_use_modules(log_converter Xml Core)
```
Comment (JNeiger) [inline logging/CMakeLists.txt:13]: This feels unrelated to this pr. Is this part of the process to save context every frame as a data logging solution or something?

File: soccer/RobotConfig.hpp [hunk 5]
```diff
@@ -78,11 +84,14 @@ class RobotConfig {
...
+    RobotStatus() {}
     RobotStatus(Configuration* config, QString prefix);
-    ~RobotStatus() {}
 
-    ConfigBool* chipper_enabled;
-    ConfigBool* kicker_enabled;
-    ConfigBool* ball_sense_enabled;
-    ConfigBool* dribbler_enabled;
+    RobotStatus(RobotStatus&&) = default;
+    RobotStatus& operator=(RobotStatus&&) = default;
```
Comment (JNeiger) [inline soccer/RobotConfig.hpp:91]: Are we sure we want to allow copies with all the unique ptrs here?

File: soccer/motion/MotionControlNode.cpp (hunk 3 above)
Comment (kylestach) [reply soccer/motion/MotionControlNode.cpp:24]: I mean, it's not really that weird. It's "address of the value at index i in the array pointed to by setpoints", which is a pretty reasonable thing to express. Either way not much we can do about it.

File: logging/CMakeLists.txt (hunk 4 above)
Comment (kylestach) [reply logging/CMakeLists.txt:13]: No, this is because the per-robot configurables got added into context, which log_converter depends upon. When this isn't here log_converter will (eventually) include context, which will include some Qt files from the Xml package, so it will fail. The "proper" fix is that `qt5_use_modules` should transitively allow the include directories (the macro is kind of broken, but it's not our code) - however, this is a reasonable workaround.

File: soccer/motion/MotionControl.cpp [hunk 6]
```diff
@@ -35,244 +30,150 @@ void MotionControl::createConfiguration(Configuration* cfg) {
...
-
-    // Clamp World Acceleration
-    auto dt = RJ::Seconds(RJ::now() - _lastCmdTime);
-    Point targetAccel = (target.vel - _lastWorldVelCmd) / dt.count();
-    targetAccel.clamp(*_max_acceleration);
-
-    target.vel = _lastWorldVelCmd + targetAccel * dt.count();
-
-    _lastWorldVelCmd = target.vel;
-    _lastCmdTime = RJ::now();
```
Comment (JNeiger) [inline soccer/motion/MotionControl.cpp:206]: Just to throw this out there, it's more correct without all the motion control hacks, so removing them is the "correct" thing to do, but it did bite us in the ass when I did that on firmware.  It might be prudent to leave them in until tuning phase, and then remove them if they're not needed.

Comment (kylestach) [reply soccer/motion/MotionControl.cpp:206]: That's true. However, the minimum-velocity hacks actually significantly (negatively) impact sim performance, and so I think they belong on the robot side of motion control regardless. I'd honestly rather optimize this part for sim anyways, considering that we are who knows how far from seeing an actual robot, and the first thing we do with new robots is motion control work where I'd rather have a clean(er) slate to work with anyway.

Acceleration _feedthrough_ is something that I think probably does belong here. I'll add it in tomorrow. Acceleration limiting is more questionable  - there's an argument to be said that it belongs on the robot, or that it belongs here, but either way I don't think it will be making much difference considering the motion planning already acceleration limits the goal (so it only matters here if feedback is large).

File: soccer/RobotConfig.hpp (hunk 5 above)
Comment (kylestach) [reply soccer/RobotConfig.hpp:91]: That's a move. Moving is okay, copying isn't.

File: soccer/motion/MotionControl.cpp (hunk 2 above)
Comment (kylestach) [reply soccer/motion/MotionControl.cpp]: Yeah, I'll add that in tomorrow. I just couldn't think of a good place for the "is joystick controlled" field to go in Context.

File: soccer/motion/MotionControl.cpp (hunk 1 above)
Comment (kylestach) [reply soccer/motion/MotionControl.cpp:129]: Fixed

File: soccer/RobotConfig.hpp (hunk 5 above)
Comment (JNeiger) [reply soccer/RobotConfig.hpp:91]: Oof, super misread that

**Review by JNeiger — APPROVED** [review]
LGTM

File: soccer/Processor.cpp [hunk 7]
```diff
@@ -121,6 +135,17 @@ Processor::~Processor() {
     // DEBUG - This is unnecessary, but lets us determine which one breaks.
     //_refereeModule.reset();
     _gameplayModule.reset();
+
+    // Put back configurables where we found them.
+    // This is kind of a hack, but if we don't do that they get destructed
+    // when Processor dies. That normally isn't a problem, but in unit tests,
+    // we create and destroy multiple instances of Processor for each test.
+    robot_config_init = std::move(_context.robot_config);
+    std::cout << robot_config_init.get() << std::endl;
```
Comment (oswinso) [inline soccer/Processor.cpp]: debug?

File: soccer/motion/MotionSetpoint.hpp [hunk 8]
```diff
@@ -1,9 +1,9 @@
 #pragma once
 // stores the outputs published by MotionControl
 struct MotionSetpoint {
-    float xvelocity;
-    float yvelocity;
-    float avelocity;
+    double xvelocity;
+    double yvelocity;
+    double avelocity;
```
Comment (oswinso) [inline soccer/motion/MotionSetpoint.hpp:6]: is there a reason for this change? Aren't floats used everywhere though?

**Review by oswinso — COMMENTED** [review]
two comments

File: soccer/motion/MotionSetpoint.hpp (hunk 8 above)
Comment (kylestach) [reply soccer/motion/MotionSetpoint.hpp:6]: I mean, yes they are in a lot of places, but that doesn't really mean they should be. There's really no reason not to use doubles. I changed some motion control code to use double, so I figured I might as well change this too.

Comment (oswinso) [reply soccer/motion/MotionSetpoint.hpp:6]: I guess we should make an issue to just straight up change all floats to doubles then

Comment (kylestach) [reply soccer/motion/MotionSetpoint.hpp:6]: Probably

**Review by oswinso — APPROVED** [review]
lgtm
